docs/tech/rfc/0019-240310-cesium-race-conditions.md

```diff
@@ -7,129 +7,833 @@
+
+### 4.1.2 Contentions with reading
+
+Reading does not modify data in the domain index nor in the file system. Therefore, it will
+not contend with other operations. However, other operations that modify the domain index
+or the file system will affect correct reads – see discussion in [4.2.2](#422-reading).
+
+### 4.1.3 Write-Delete contention
+
+The deleter does not change data on the file system, so we only need to consider the
+concurrent operations on domain the index by the writer and the deleter.
+We will discuss the two cases: either the writer may write data to the range
+requested to be deleted, or it may not.
+
+In the case that the writer does not write to the deleted time range, the mutex on the
+domain index is sufficient to handle the concurrent writes as the writer and deleter are
+necessarily writing to different parts of the domain index, so there will be no conflicts.
+
+The case where the writer may write to the deleted time range, however, will cause domain index
+conflicts: if the write operation happens first, then a new domain will be inserted and
+then deleted from the deletion operation. However, if the deletion operation happens
+first, then there will be a chunk of written data in the time range despite the delete call.
+This is not the behaviour we want. So we use the controller to disallow all deletions
+to a time range that a writer may write to.
+
+### 4.1.4 Write-Garbage Collection Contention
+
+Both the writer and the garbage collector modify the underlying data as well as the index.
+In addition, since Garbage Collection is fundamentally a secondary operation (compared
+to reads, writes, and deletions), it should not affect those operations – it is unrealistic
+to disallow reads and writes while garbage collecting.
+
+There are no major conflicts between a writer and Garbage Collection:
+Garbage Collection is strictly only run on a file that is oversize and has no writers
+currently writing to it. This means that no writer can write to a file currently being
+garbage collected.
+
+In the domain index, since Garbage Collection only modifies pointers on the file being
+collected, it will not change any pointers being written to by the writer, as it is
+necessarily not on the file, for reasons mentioned above.
+
+These conflicts are taken care of by the file controller and the mutex, and
+there are no additional actions needed.
+
+### 4.1.5 Delete-Garbage Collection contention
+
+The deleter does not modify the data in the file system, so we will only discuss the
+contentions in the domain index.
+Unlike the writer, the deleter may perform operations on the pointers currently being
+garbage collected.
+
+The conflict arises when both entities attempt to change the offset of a file: delete
+must change this to account for deletion after the start of a domain to only take in
+a part of the data; Garbage Collection must change this to store the new position of the data in the
```

Comment:
Don't use the phrase "must change this". Replace "this" with "the offset" to be more clear.